src/common/product_models/heatmap_common_models.py

```diff
@@ -198,3 +199,103 @@ class CellDetail(BaseModel):
+
+
+def transform_heatmap_row_cells(data: dict[str, Any]):
+    """
+    Transform, in place, the cells structure in a heatmap row to a new structure.
+
+    The new structure is a set of keys and values where the keys are constructed from the old structure.
+    new structured key format: <HEATMAP_COL_PREFIX>_<col_id>_<FIELD_HEATMAP_CELL_ID|FIELD_HEATMAP_CELL_VALUEl>
+
+    e.g.
+
+    The old structure:
+    "cells": [
+        {
+            "cell_id": "cell_0",
+            "col_id": "0",
+            "val": 0.0
+        },
+        {
+            "cell_id": "cell_1",
+            "col_id": "1",
+            "val": 1.0
+        }
+    ]
+
+    The new structure:
+    "col_0_cell_id": "cell_0",
+    "col_0_val": 0.0,
+    "col_1_cell_id": "cell_1",
+    "col_1_val": 1.0
+
+    """
+
+    # Iterate over the 'cells' structure and remove them from the data structure while constructing the new structure
+    for cell in data.pop(FIELD_HEATMAP_ROW_CELLS):
+        col_id = cell.get(FIELD_HEATMAP_COL_ID)
+
+        # Construct keys and values for the new structure
+        cell_id_key = f"{HEATMAP_COL_PREFIX}{HEATMAP_COL_SEPARATOR}{col_id}{HEATMAP_COL_SEPARATOR}{FIELD_HEATMAP_CELL_ID}"
+        cell_val_key = f"{HEATMAP_COL_PREFIX}{HEATMAP_COL_SEPARATOR}{col_id}{HEATMAP_COL_SEPARATOR}{FIELD_HEATMAP_CELL_VALUE}"
+
+        # Add the new keys and values to the data structure
+        data[cell_id_key] = cell.get(FIELD_HEATMAP_CELL_ID)
+        data[cell_val_key] = cell.get(FIELD_HEATMAP_CELL_VALUE)
+
+
+def revert_transformed_heatmap_row_cells(data: dict[str, Any]):
+    """
+    Revert, in place, the transformation of heatmap row cells.
+    Input key format: <HEATMAP_COL_PREFIX>_<col_id>_<FIELD_HEATMAP_CELL_ID|FIELD_HEATMAP_CELL_VALUEl>
+
+    The structure (input) to be reverted:
+    "col_0_cell_id": "cell_0",
+    "col_0_val": 0.0,
+    "col_1_cell_id": "cell_1",
+    "col_1_val": 1.0
+
+    The resulting structure:
+    "cells": [
+        {
+            "cell_id": "cell_0",
+            "col_id": "0",
+            "val": 0.0
+        },
+        {
+            "cell_id": "cell_1",
+            "col_id": "1",
+            "val": 1.0
+        }
+    ]
+    """
+
+    # Get the keys that need to be reconstructed
+    keys_to_remove = [key for key in data.keys() if key.startswith(HEATMAP_COL_PREFIX) and
+                      (key.endswith(FIELD_HEATMAP_CELL_ID) or key.endswith(FIELD_HEATMAP_CELL_VALUE))]
+
+    cells_dict = {}
+    for key in keys_to_remove:
+        parts = key.split(HEATMAP_COL_SEPARATOR)
+        col_id = parts[1]
+        col_type = HEATMAP_COL_SEPARATOR.join(parts[2:])
+        if not col_id.isdigit():
+            raise ValueError(f"Column ID '{col_id}' is not an integer.")
+
+        if col_type not in [FIELD_HEATMAP_CELL_ID, FIELD_HEATMAP_CELL_VALUE]:
+            raise ValueError(f'Unexpected column type: {col_type}')
+
+        cells_dict.setdefault(col_id, {FIELD_HEATMAP_CELL_ID: None, FIELD_HEATMAP_CELL_VALUE: None})
```

Comment:
I'm not sure I understand this line. If there's no cell id or value that seems like there's an error in the data

Reply:
this was just to ensure cells_dict has the <col_id> key. 
At the beginning cells_dict doesn't have col_id, and it creates {col_id: {'cell_id': None, 'val': None}}. Right after it, it sets one of cell_id or val value. The next time, since col_id exists, setdefault has no affect, and it sets the other value.

Reply:
Why not just do
```
cells_dict.setdefault(col_id, {})
```

Reply:
no specific reason. I want `cells_dict[col_id][col_type]` to fail if `col_type` is missing. But I guess that won't happen either since I did check col_type earlier.

Reply:
It should now cause a key error down below if one of the 2 col types are missing

Reply:
:+1:

Reply:
Setting it to empty dict is actually better since `cell_data[FIELD_HEATMAP_CELL_VALUE]` will fail later.

Reply:
yeah, that's what I meant